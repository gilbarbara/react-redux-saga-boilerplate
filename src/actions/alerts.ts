import { ReactNode } from 'react';
import { uuid } from '@gilbarbara/helpers';
import { createAction } from '@reduxjs/toolkit';

import { actionPayload } from '~/modules/helpers';

import { ActionTypes } from '~/literals';

import { ShowAlertOptions } from '~/types';

export const hideAlert = createAction<string>(ActionTypes.ALERTS_HIDE);

export const showAlert = createAction(
  ActionTypes.ALERTS_SHOW,
  (message: ReactNode, options: ShowAlertOptions) => {
    const timeout = options.type === 'error' ? 0 : 5;

    return actionPayload({
      id: options.id || uuid(),
      icon: options.icon || 'info-o',
      message,
      position: options.position || 'bottom-right',
      type: options.type || 'neutral',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    });
  },
);
