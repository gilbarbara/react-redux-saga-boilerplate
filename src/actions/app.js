// @flow
/**
 * @module Actions/App
 * @desc App Actions
 */

import uuid from 'uuid/v4';
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export { goBack, go, push, replace } from 'modules/history';

export const { hideAlert, showAlert, switchMenu } = createActions({
  [ActionTypes.SWITCH_MENU]: (query: string) => ({ query }),
  [ActionTypes.HIDE_ALERT]: (id: string) => ({ id }),
  [ActionTypes.SHOW_ALERT]: (message: string, options: Object) => {
    const timeout = options.variant === 'danger' ? 0 : 5;

    return {
      id: options.id || uuid(),
      icon: options.icon,
      message,
      position: options.position || 'bottom-right',
      variant: options.variant || 'dark',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    };
  },
});
