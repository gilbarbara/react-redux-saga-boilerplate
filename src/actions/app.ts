/**
 * @module Actions/App
 * @desc App Actions
 */
import React from 'react';
import { nanoid } from 'nanoid';

import { createAction } from 'modules/helpers';

import { ActionTypes } from 'literals';

import { ShowAlertOptions } from 'types';

export { go, goBack, push, replace } from 'modules/history';

export const hideAlert = createAction(ActionTypes.HIDE_ALERT, (id: string) => ({ id }));

export const showAlert = createAction(
  ActionTypes.SHOW_ALERT,
  (message: React.ReactNode, options: ShowAlertOptions) => {
    const timeout = options.variant === 'danger' ? 0 : 5;

    return {
      id: options.id || nanoid(),
      icon: options.icon,
      message,
      position: options.position || 'bottom-right',
      variant: options.variant || 'dark',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    };
  },
);
