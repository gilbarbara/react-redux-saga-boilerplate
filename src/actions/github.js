// @flow
/**
 * @module Actions/User
 * @desc User Actions
 */
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export const { githubGetRepos: getRepos } = createActions({
  [ActionTypes.GITHUB_GET_REPOS]: (query: string) => ({ query }),
});
