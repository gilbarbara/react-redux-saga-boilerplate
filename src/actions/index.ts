export * from './app';
export * from './github';
export * from './user';
