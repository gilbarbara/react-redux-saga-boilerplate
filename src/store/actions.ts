export { alertHide, alertShow } from './slices/alerts';
export { setAppOptions } from './slices/app';
export { getRepos, getReposFailure, getReposSuccess } from './slices/github';
export { login, loginSuccess, logOut, logOutSuccess } from './slices/user';
