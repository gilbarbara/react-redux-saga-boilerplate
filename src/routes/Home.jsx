import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

import config from 'config';
import { login } from 'actions/index';

import { Button, Container, Text, utils } from 'styled-minimal';
import Background from 'components/Background';
import Icon from 'components/Icon';
import Logo from 'components/Logo';

const { spacer } = utils;

const HomeWrapper = styled(Background)``;

const HomeContainer = styled(Container)`
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: center;
  min-height: 100vh;
`;

const Header = styled.div`
  margin-bottom: ${spacer(3)};
  text-align: center;

  svg {
    width: 20rem;
  }
`;

const Heading = styled.h1`
  color: #fff;
  font-size: 3.5rem;
  line-height: 1.4;
  margin-bottom: ${spacer(3)};
  margin-top: 0;
  text-align: center;

  ${utils.responsive({
    lg: `
      font-size: 4rem;
    `,
  })}
`;

export class Home extends React.PureComponent {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired,
  };

  handleClickLogin = () => {
    const { dispatch } = this.props;

    dispatch(login());
  };

  render() {
    const { user } = this.props;

    return (
      <HomeWrapper key="Home">
        <HomeContainer verticalPadding>
          <Header>
            <Logo />
          </Header>
          <Heading>{config.description}</Heading>
          <Button
            animate={user.status === 'running'}
            onClick={this.handleClickLogin}
            size="xl"
            textTransform="uppercase"
          >
            <Icon name="sign-in" />
            <Text ml={2}>Start</Text>
          </Button>
        </HomeContainer>
      </HomeWrapper>
    );
  }
}

/* istanbul ignore next */
function mapStateToProps(state) {
  return { user: state.user };
}

export default connect(mapStateToProps)(Home);
