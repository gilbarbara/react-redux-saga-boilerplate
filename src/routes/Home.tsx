import { useDispatch, useSelector } from 'react-redux';
import styled from '@emotion/styled';
import { Box, Button, Container, Icon, responsive, Text } from '@gilbarbara/components';

import { name } from '~/config';

import theme from '~/modules/theme';

import { login } from '~/actions';
import { STATUS } from '~/literals';

import Background from '~/components/Background';
import Logo from '~/components/Logo';

import { RootState } from '~/types';

const Header = styled.div`
  margin-bottom: ${theme.spacing.lg};
  text-align: center;

  svg {
    height: 100px;
    width: auto;

    ${responsive({
      lg: {
        height: '180px',
      },
    })};
  }
`;

const Heading = styled.h1`
  color: #fff;
  font-size: 35px;
  line-height: 1.4;
  margin-bottom: ${theme.spacing.lg};
  margin-top: 0;
  text-align: center;

  ${responsive({
    lg: {
      fontSize: '40px',
    },
  })};
`;

function Home() {
  const dispatch = useDispatch();
  const status = useSelector<RootState>(({ user }) => user.status);

  const handleClickLogin = () => {
    dispatch(login());
  };

  return (
    <Background key="Home" data-component-name="Home">
      <Container fullScreen justify="center">
        <Box textAlign="center">
          <Header>
            <Logo />
          </Header>
          <Heading>{name}</Heading>
          <Button
            bg="white"
            busy={status === STATUS.RUNNING}
            data-component-name="Start"
            onClick={handleClickLogin}
            size="lg"
          >
            <Icon mr="xs" name="star" size={24} />
            <Text size="lg">Start</Text>
          </Button>
        </Box>
      </Container>
    </Background>
  );
}

export default Home;
