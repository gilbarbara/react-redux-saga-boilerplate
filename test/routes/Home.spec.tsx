import React from 'react';
import { fireEvent, render, screen } from 'test-utils';

import { login } from '~/actions';

import Home from '~/routes/Home';

const mockDispatch = vi.fn();

describe('Home', () => {
  it('should render properly', () => {
    render(<Home />);
    expect(screen.getByTestId('Home')).toMatchSnapshot();
  });

  it('should handle clicks', () => {
    render(<Home />, { mockDispatch });
    fireEvent.click(screen.getByTestId('Start'));

    expect(mockDispatch).toHaveBeenCalledWith(login());
  });
});
