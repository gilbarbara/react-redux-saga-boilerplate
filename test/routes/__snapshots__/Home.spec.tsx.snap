// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home > should render properly 1`] = `
<div
  class="css-146gbh6"
  data-component-name="Home"
>
  <div
    class="css-1ym6d5y"
    data-component-name="Container"
  >
    <div
      class="css-167s73c"
      data-component-name="Box"
    >
      <div
        class="css-1llpsq5"
      >
        <div
          class="css-11mttn6"
        >
          <svg
            src="/media/brand/icon.svg"
          />
        </div>
      </div>
      <h1
        class="css-2kvgnr"
      >
        React Redux Saga Boilerplate
      </h1>
      <button
        class="css-3qd3v6"
        data-component-name="Start"
        type="button"
      >
        <span
          class="css-1k4akb"
          data-component-name="Icon"
          name="star"
        >
          <svg
            height="24"
            src="https://files.gilbarbara.dev/icons/star.svg"
            title="star"
            width="24"
          />
        </span>
        <span
          class="css-iz9y3l"
          data-component-name="Text"
        >
          Start
        </span>
      </button>
    </div>
  </div>
</div>
`;
